File: morpheus/io/deserializers.py
# SPDX-FileCopyrightText: Copyright (c) 2021-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""DataFrame deserializers."""

import io
import typing

import numpy as np
import pandas as pd

if typing.TYPE_CHECKING:
    import cudf

from morpheus.common import FileTypes
from morpheus.common import determine_file_type
from morpheus.common import read_file_to_df as read_file_to_df_cpp
from morpheus.config import CppConfig
from morpheus.io.utils import filter_null_data
from morpheus.utils.type_aliases import DataFrameType


def _read_file_to_df_py(*,
                        file_name: typing.Union[str, io.IOBase],
                        file_type: FileTypes,
                        parser_kwargs: dict,
                        df_type: typing.Literal["cudf", "pandas"]) -> DataFrameType:
    if (parser_kwargs is None):
        parser_kwargs = {}

    mode = file_type

    if (mode == FileTypes.Auto):
        # The DFPFileToDataFrameStage passes an instance of an fsspec file opener instead of a filename to this method.
        # The opener objects are subclasses of io.IOBase, which avoids introducing fsspec to this part of the API
        if (isinstance(file_name, io.IOBase)):
            if (hasattr(file_name, 'path')):  # This attr is not in the base
                filepath = file_name.path
            else:
                raise ValueError("Unable to determine file type from instance of io.IOBase,"
                                 " set `file_type` to a value other than Auto")
        else:
            filepath = file_name
        mode = determine_file_type(filepath)

    # Special args for JSON
    kwargs = {}
    if (mode == FileTypes.JSON):
        kwargs["lines"] = True

    # Update with any args set by the user. User values overwrite defaults
    kwargs.update(parser_kwargs)

    if df_type == "cudf":
        import cudf
        df_class = cudf
    else:
        df_class = pd

    df = None
    if (mode == FileTypes.JSON):
        df = df_class.read_json(file_name, **kwargs)

    elif (mode == FileTypes.CSV):
        df: DataFrameType = df_class.read_csv(file_name, **kwargs)

        if (len(df.columns) > 1 and df.columns[0] == "Unnamed: 0" and df.iloc[:, 0].dtype == np.dtype(int)):
            df.set_index("Unnamed: 0", drop=True, inplace=True)
            df.index.name = ""
            df.sort_index(inplace=True)

    elif (mode == FileTypes.PARQUET):
        df = df_class.read_parquet(file_name, **kwargs)

    else:
        assert False, f"Unsupported file type mode: {mode}"

    assert df is not None

    return df


def read_file_to_df(file_name: typing.Union[str, io.IOBase],
                    file_type: FileTypes = FileTypes.Auto,
                    parser_kwargs: dict = None,
                    filter_nulls: bool = True,
                    filter_null_columns: list[str] | str = 'data',
                    df_type: typing.Literal["cudf", "pandas"] = "pandas") -> DataFrameType:
    """
    Reads a file into a dataframe and performs any of the necessary cleanup.

    Parameters
    ----------
    file_name : str
        File to read.
    file_type : `morpheus.common.FileTypes`
        Type of file. Leave as Auto to determine from the extension.
    parser_kwargs : dict, optional
        Any argument to pass onto the parse, by default {}. Ignored when C++ execution is enabled and `df_type="cudf"`
    filter_nulls : bool, optional
        Whether to filter null rows after loading, by default True.
    filter_null_columns : list[str]|str, default = 'data'
        Column or columns to filter null values from. Ignored when `filter_null` is False.
    df_type : typing.Literal[, optional
        What type of parser to use. Options are 'cudf' and 'pandas', by default "pandas".

    Returns
    -------
    DataFrameType
        A parsed DataFrame.
    """

    # The C++ reader only supports cudf dataframes
    if (CppConfig.get_should_use_cpp() and df_type == "cudf"):
        df = read_file_to_df_cpp(file_name, file_type)
    else:
        df = _read_file_to_df_py(file_name=file_name, file_type=file_type, parser_kwargs=parser_kwargs, df_type=df_type)

    if (filter_nulls):
        if isinstance(filter_null_columns, str):
            filter_null_columns = [filter_null_columns]

        for col in filter_null_columns:
            df = filter_null_data(df, column_name=col)

    return df

File: python/morpheus/morpheus/_lib/src/io/deserializers.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "morpheus/io/deserializers.hpp"

#include "morpheus/utilities/cudf_util.hpp"  // for CudfHelper
#include "morpheus/utilities/stage_util.hpp"
#include "morpheus/utilities/string_util.hpp"
#include "morpheus/utilities/table_util.hpp"  // for get_column_names

#include <cudf/column/column.hpp>
#include <cudf/io/csv.hpp>
#include <cudf/io/json.hpp>
#include <cudf/io/parquet.hpp>
#include <cudf/table/table.hpp>  // IWYU pragma: keep
#include <cudf/types.hpp>        // for cudf::type_id
#include <pybind11/pybind11.h>   // IWYU pragma: keep

#include <memory>
#include <regex>
#include <sstream>
#include <stdexcept>
#include <utility>
// We're already including pybind11.h, and including only gil.h as IWYU suggests yields an undefined symbol error
// IWYU pragma: no_include <pybind11/gil.h>

namespace {
const std::regex INDEX_REGEX(R"(^\s*(unnamed: 0|id)\s*$)",
                             std::regex_constants::ECMAScript | std::regex_constants::icase);

const std::regex UNNAMED_REGEX(R"(^\s*unnamed: 0\s*$)", std::regex_constants::ECMAScript | std::regex_constants::icase);
}  // namespace

namespace morpheus {

std::vector<std::string> get_column_names_from_table(const cudf::io::table_with_metadata& table)
{
    return foreach_map(table.metadata.schema_info, [](auto schema) {
        return schema.name;
    });
}

cudf::io::table_with_metadata load_table_from_file(const std::string& filename,
                                                   FileTypes file_type,
                                                   std::optional<bool> json_lines)
{
    if (file_type == FileTypes::Auto)
    {
        file_type = determine_file_type(filename);  // throws if it is unable to determine the type
    }

    cudf::io::table_with_metadata table;

    switch (file_type)
    {
    case FileTypes::JSON: {
        auto options =
            cudf::io::json_reader_options::builder(cudf::io::source_info{filename}).lines(json_lines.value_or(true));
        table = cudf::io::read_json(options.build());
        break;
    }
    case FileTypes::CSV: {
        auto options = cudf::io::csv_reader_options::builder(cudf::io::source_info{filename});
        table        = cudf::io::read_csv(options.build());
        break;
    }
    case FileTypes::PARQUET: {
        auto options = cudf::io::parquet_reader_options::builder(cudf::io::source_info{filename});
        table        = cudf::io::read_parquet(options.build());
        break;
    }
    case FileTypes::Auto:
    default:
        throw std::logic_error(MORPHEUS_CONCAT_STR("Unsupported filetype: " << file_type));
    }

    if (!table.tbl)
    {
        throw std::runtime_error(MORPHEUS_CONCAT_STR("Failed to load file '" << filename << "' as type " << file_type));
    }

    return table;
}

pybind11::object read_file_to_df(const std::string& filename, FileTypes file_type)
{
    auto table          = load_table_from_file(filename, file_type);
    int index_col_count = prepare_df_index(table);

    pybind11::gil_scoped_acquire gil;
    return CudfHelper::table_from_table_with_metadata(std::move(table), index_col_count);
}

int get_index_col_count(const cudf::io::table_with_metadata& data_table)
{
    int index_col_count = 0;

    std::vector<std::string> names = CuDFTableUtil::get_column_names(data_table);

    // Check if we have a first column with INT64 data type
    if (names.size() >= 1 && data_table.tbl->get_column(0).type().id() == cudf::type_id::INT64)
    {
        // Get the column name
        const auto& col_name = names[0];

        // Check it against some common terms
        if (std::regex_search(col_name, INDEX_REGEX))
        {
            index_col_count = 1;
        }
    }

    return index_col_count;
}

int prepare_df_index(cudf::io::table_with_metadata& data_table)
{
    const int index_col_count = get_index_col_count(data_table);

    if (index_col_count > 0)
    {
        auto& col_name = data_table.metadata.schema_info[0].name;

        // Also, if its the hideous 'Unnamed: 0', then just use an empty string
        if (std::regex_search(col_name, UNNAMED_REGEX))
        {
            col_name.clear();
        }
    }

    return index_col_count;
}

}  // namespace morpheus
